### PR TITLE
READMEs

### PR DESCRIPTION
**Pull request by Jolanrensen**
Fixes https://github.com/Kotlin/dataframe/issues/844

Please @koperagen and @zaleslaw check the READMEs contents for correctness.
I dove into each module and folder and wrote what it contained to my best understanding, but if I made any mistakes, let me know :)

Let's hope this helps us and contributors to no longer get lost in the project :)

Spawns two new issues that can be resolved later:
- https://github.com/Kotlin/dataframe/issues/898
- https://github.com/Kotlin/dataframe/issues/899

### CODE REVIEW

File: .github/README.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
+## GitHub Actions
+
+While publishing and testing of the library is handled by [JetBrains TeamCity](https://www.jetbrains.com/teamcity/),
+there are some CI operations that are handled by GitHub actions instead.
+
+### Publishing Docs
+
+The building of the documentation website in [docs](../docs), and the publishing of it along
```
Comment (zaleslaw) [inline .github/README.md:8]: Let's add an info that publishing fully replaces previous documentation (we have no versioning)

File: core/README.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
+## :core
+
+This is the core of the library, published as the `dataframe-core` package.
+It contains the DataFrame API and its implementation, as well as plenty of JUnit tests.
+
+I/O operations are split off into other modules, like [:dataframe-excel](../dataframe-excel),
+or [:dataframe-jdbc](../dataframe-jdbc), however, this is has not happened yet for all operations
+(see [Issue #100](https://github.com/Kotlin/dataframe/issues/100)).
+
+Operations likely to be split off in the future include:
```
Comment (zaleslaw) [inline core/README.md:10]: probably better to remove about future and fix the current state and say, that these data sources are part of Core

File: docs/README.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
+## The Documentation Website
+
+This folder holds the source code of our documentation website:
+[kotlin.github.io/dataframe](https://kotlin.github.io/dataframe).
+
+It's built using [WriterSide](https://www.jetbrains.com/writerside/) and published
+by a [Github Action](../.github/workflows/main.yml).
+The file structure mirrors the default WriterSide structure with some small exceptions.
+
+Images all README files can be stored in [docs/imgs](./imgs).
```
Comment (zaleslaw) [inline docs/README.md:10]: Let's mention the file with page structure, and the fact that all new pages should be inserted into it

File: docs/README.md [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+which hooks into [the TestBase class](../core/src/test/kotlin/org/jetbrains/kotlinx/dataframe/samples/api/TestBase.kt) and
+retrieves the intermediate DataFrame expressions thanks to
+our "explainer" compiler plugin [:plugins:expressions-converter](../plugins/expressions-converter).
+
+We can also generate "normal" DataFrame samples for the website. This can be done using the
+[OtherSamples class](../core/src/test/kotlin/org/jetbrains/kotlinx/dataframe/samples/api/OtherSamples.kt). Generated
+HTML files will be stored in [docs/StardustDocs/snippets/manual](./StardustDocs/snippets/manual).
+
+### KDoc Preprocesor
+KDocs can also be exported to HTML, for them to be reused on the website.
```
Comment (zaleslaw) [inline docs/README.md:26]: Is it copypaste from core/Readme?

**Review by zaleslaw — CHANGES_REQUESTED** [review]
Wow, great job, I've discovered one plugin which I never used/heard!

**Comment by Jolanrensen** [comment]
TODO Add readme for https://github.com/Kotlin/dataframe/pull/916 when it's merged

File: docs/README.md (hunk 4 above)
Comment (Jolanrensen) [reply docs/README.md:26]: yes, well, only the part relevant to the docs

File: core/README.md [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+It contains the DataFrame API and its implementation, as well as plenty of JUnit tests.
+
+I/O operations are split off into other modules, like [:dataframe-excel](../dataframe-excel),
+or [:dataframe-jdbc](../dataframe-jdbc), however, this is has not happened yet for all operations
+(see [Issue #100](https://github.com/Kotlin/dataframe/issues/100)).
+
+At the moment, these data sources are still part of the `:core` module:
+
+- csv/tsv
+- html
```
Comment (koperagen) [inline core/README.md:13]: Not data source? Same with jupyter

File: dataframe-excel/README.md [hunk 6]
```diff
@@ -0,0 +1,8 @@
+## :dataframe-excel
+
+This module, published as `dataframe-excel`, contains all logic and tests for DataFrame to be able to work with
+Excel files.
```
Comment (koperagen) [inline dataframe-excel/README.md:4]: I usually mention `xls` and `xlsx` specifically

File: dataframe-openapi-generator/README.md [hunk 7]
```diff
@@ -0,0 +1,13 @@
+## :dataframe-openapi
```
Comment (koperagen) [inline dataframe-openapi-generator/README.md:1]: title should be dataframe-openapi-generator too

File: dataframe-openapi-generator/README.md [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+
+This module, published as `dataframe-openapi-generator` contains all logic and tests for DataFrame to be able to import
+OpenAPI specifications as auto-generated data schemas. This module is a sister module to
+[`dataframe-openapi`](../dataframe-openapi):
+
+- `dataframe-openapi-generator` is used as a dependency of the Gradle plugin and Jupyter plugin to be able to generate
```
Comment (koperagen) [inline dataframe-openapi-generator/README.md:7]: KSP can import schemas too, right?

File: docs/README.md [hunk 9]
```diff
@@ -0,0 +1,40 @@
+## The Documentation Website
+
+This folder holds the source code of our documentation website:
+[kotlin.github.io/dataframe](https://kotlin.github.io/dataframe).
+
+It's built using [WriterSide](https://www.jetbrains.com/writerside/) and published
```
Comment (koperagen) [inline docs/README.md:6]: Proper casing here is Writside

File: plugins/README.md [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+
+This folder holds all our Gradle- and Kotlin Compiler plugins:
+
+### [:plugins:dataframe-gradle-plugin](./dataframe-gradle-plugin)
+The Gradle plugin for DataFrame that can generate data schemas from a data sample using the simple
+task `dataframes {}`. It uses [:plugins:symbol-processor](./symbol-processor) to generate column accessors.
+
+### [:plugins:symbol-processor](./symbol-processor)
+The KSP plugin that can generate data schemas from a data sample using the `@file:ImportDataSchema` annotation.
+It is also used to detect (both manually written- or generated) `@DataSchema` annotated classes/interfaces to generate
```
Comment (koperagen) [inline plugins/README.md:11]: I'd rephrase to just "Generated extension properties for `DataSchema` annotated classes and interfaces, omit "detect" part

File: plugins/README.md [hunk 11]
```diff
@@ -0,0 +1,28 @@
+## :plugins
+
+This folder holds all our Gradle- and Kotlin Compiler plugins:
+
+### [:plugins:dataframe-gradle-plugin](./dataframe-gradle-plugin)
+The Gradle plugin for DataFrame that can generate data schemas from a data sample using the simple
+task `dataframes {}`. It uses [:plugins:symbol-processor](./symbol-processor) to generate column accessors.
```
Comment (koperagen) [inline plugins/README.md:7]: > It uses [:plugins:symbol-processor](./symbol-processor) to generate column accessors.

It applies KSP plugin to your build and symbol-processor for easier configuration

File: plugins/README.md [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+
+### [:plugins:symbol-processor](./symbol-processor)
+The KSP plugin that can generate data schemas from a data sample using the `@file:ImportDataSchema` annotation.
```
Comment (koperagen) [inline plugins/README.md:10]: suggestion:
This is an alternative to `dataframes { }` syntax from Gradle, it does the same thing but declared closer to your source code

File: plugins/README.md (hunk 11 above)
Comment (koperagen) [reply plugins/README.md:7]: Oh, ok, exactly what you wrote in the next file :o

File: plugins/kotlin-dataframe/README.md [hunk 13]
```diff
@@ -0,0 +1,10 @@
+## :plugins:kotlin-dataframe
+
+This module, published as "compiler-plugin-all", holds the Kotlin 2.x Compiler plugin of DataFrame.
+
+A [work-in-progress](https://github.com/Kotlin/dataframe/issues/704)
+plugin for your Gradle project that can generate on-the-fly column accessors for the compiler and IDE even without
```
Comment (koperagen) [inline plugins/kotlin-dataframe/README.md:6]: Not sure about "Gradle project" detail. It will work in notebooks too in not-so-distant future

File: plugins/kotlin-dataframe/README.md [hunk 14]
```diff
@@ -0,0 +1,10 @@
...
+having to provide data schemas!
+
+Details of how to build and run this for yourself will follow, but there is a
```
Comment (koperagen) [inline plugins/kotlin-dataframe/README.md:9]: What do you mean by build? publish to maven local and use local build?

File: generator/README.md [hunk 15]
```diff
@@ -0,0 +1,14 @@
...
+[:core](../core) with the `generateKeywordsSrc` task.
+
+This task, generates three enum classes: `HardKeywords`, `ModifierKeywords`, and `SoftKeywords`.
+These enums together contain all restricted Kotlin keywords to be taken into account when generating our own
+code in Notebooks or any of our [plugins](../plugins). Words like "package", "fun", "suspend", etc...
+
+As the Kotlin language can change over time, this task ensures that any changes to the language
+will be reflected in our code generation.
+
+This module can probably be moved under [:plugins](../plugins):
```
Comment (koperagen) [inline generator/README.md:13]: Do you want to keep this line?

File: dataframe-openapi-generator/README.md (hunk 8 above)
Comment (Jolanrensen) [reply dataframe-openapi-generator/README.md:7]: yes definitely, but in the documentation we don't explicitly name "ksp" to be separate from the gradle plugin, right? Even I am not entirely sure about the difference because the gradle plugin also uses ksp to generate the sources right?

File: core/README.md (hunk 5 above)
Comment (Jolanrensen) [reply core/README.md:13]: true

Comment (Jolanrensen) [reply core/README.md:13]: let's call them "integrations"

File: dataframe-excel/README.md (hunk 6 above)
Comment (Jolanrensen) [reply dataframe-excel/README.md:4]: good idea, you can also produce these filetypes with google sheets for instance

File: plugins/README.md (hunk 12 above)
Comment (Jolanrensen) [reply plugins/README.md:10]: sure, but I'll put it in the symbol-processor/README.md. This is just a summary after all

File: plugins/kotlin-dataframe/README.md (hunk 13 above)
Comment (Jolanrensen) [reply plugins/kotlin-dataframe/README.md:6]: fair, it might even work in maven, won't it?

File: plugins/kotlin-dataframe/README.md (hunk 14 above)
Comment (Jolanrensen) [reply plugins/kotlin-dataframe/README.md:9]: I think I meant "details of how to use it in your own project"... but I'm not sure, I'll rephrase it

File: generator/README.md (hunk 15 above)
Comment (Jolanrensen) [reply generator/README.md:13]: Hmm I don't think it would hurt. Informs people that it is a plugin and likely belongs there but we just haven't come around to moving it yet.
